src/app/qan/profile-details/components/explain/explain.component.ts

```diff
@@ -0,0 +1,107 @@
+import { Component, OnDestroy, OnInit } from '@angular/core';
+import { ActionsService } from '../../../../pmm-api-services/services/actions.service';
+import { interval, of, Subscription } from 'rxjs';
+import { catchError, map, startWith, switchMap, take } from 'rxjs/operators';
+import { ObjectDetails, QanProfileService } from '../../../profile/qan-profile.service';
+import { ObjectDetailsService } from '../../../../pmm-api-services/services/object-details.service';
+
+@Component({
+  selector: 'app-explain',
+  templateUrl: './explain.component.html',
+  styleUrls: ['./explain.component.css']
+})
+export class ExplainComponent implements OnInit, OnDestroy {
+  private example$: Subscription;
+  private defaultExample$: Subscription;
+  private classicStart$: Subscription;
+  private jsonStart$: Subscription;
+
+  public currentDetails: ObjectDetails;
+  public classicOutput: string;
+  public jsonOutput: string;
+  public visualOutput: string;
+
+  constructor(
+    private actionsService: ActionsService,
+    protected objectDetailsService: ObjectDetailsService,
+    protected qanProfileService: QanProfileService,
+  ) {
+    this.currentDetails = this.qanProfileService.getCurrentDetails;
+    this.example$ = this.qanProfileService.getProfileInfo.details.pipe(
+      switchMap(parsedParams => this.getExample(parsedParams)))
+      .subscribe(
+        response => this.startExplainActions(response[0])
+      );
+    this.defaultExample$ = this.getExample(this.currentDetails)
+      .pipe(take(1))
+      .subscribe(response => this.startExplainActions(response[0]))
+  }
+
+  ngOnInit() {
+  }
+
+  private startExplainActions(value) {
+    this.startClassic(value);
+    this.startJson(value);
+  }
+
+  private startClassic(value) {
+    this.classicStart$ = this.actionsService.StartMySQLExplainAction({
+      service_id: value.service_id,
+      query: value.example,
+      database: value.schema,
+    }).subscribe(item => {
+      const explainResultObservable = interval(1000)
+        .pipe(
+          startWith(0),
+          switchMap(() => this.actionsService.GetAction({ action_id: item.action_id }))
+        )
+        .subscribe(res => {
+          if (res.done) {
+            this.classicOutput = res.output;
+            if (explainResultObservable) {
+              explainResultObservable.unsubscribe()
+            }
+          }
+        });
+    });
+  }
+
+  private startJson(value) {
+    this.jsonStart$ = this.actionsService.StartMySQLExplainJSONAction({
+      service_id: value.service_id,
+      query: value.example,
+      database: value.schema,
+    }).subscribe(item => {
+      const explainResultObservable = interval(1000)
```

Comment:
Is it necessary to use interval here?

Reply:
yes, we should pull data from backend each second until we get `done` flag.